File: bindings/kotlin/src/main/kotlin/com/elabify/core/Hex.kt
// Hex encoding helpers. Cross-binding-equivalent with @elabify/core's
// src/hex.ts (lowercase hex, no 0x prefix on output, tolerant of 0x prefix
// on input).

package com.elabify.core

import java.text.Normalizer

internal fun bytesToHex(bytes: ByteArray): String {
    val out = StringBuilder(bytes.size * 2)
    for (b in bytes) {
        val v = b.toInt() and 0xff
        out.append(HEX_ALPHABET[v ushr 4])
        out.append(HEX_ALPHABET[v and 0x0f])
    }
    return out.toString()
}

internal fun hexToBytes(hexIn: String): ByteArray {
    var hex = hexIn
    if (hex.startsWith("0x") || hex.startsWith("0X")) hex = hex.substring(2)
    require(hex.length % 2 == 0) { "hexToBytes: hex string must have even length, got ${hex.length}" }
    val out = ByteArray(hex.length / 2)
    for (i in out.indices) {
        val hi = digit(hex[i * 2])
        val lo = digit(hex[i * 2 + 1])
        out[i] = ((hi shl 4) or lo).toByte()
    }
    return out
}

private val HEX_ALPHABET = "0123456789abcdef".toCharArray()

private fun digit(c: Char): Int {
    return when (c) {
        in '0'..'9' -> c - '0'
        in 'a'..'f' -> c - 'a' + 10
        in 'A'..'F' -> c - 'A' + 10
        else -> error("non-hex character: $c")
    }
}

/** UTF-8 bytes of the NFC-normalized form of [s]. */
internal fun utf8Bytes(s: String): ByteArray {
    return Normalizer.normalize(s, Normalizer.Form.NFC).toByteArray(Charsets.UTF_8)
}

File: bindings/swift/Sources/ElabifyCore/Hex.swift
// Hex encoding helpers. Cross-binding-equivalent with @elabify/core's
// src/hex.ts (lowercase hex, no 0x prefix, accepts upper- or lower-case
// on decode).

import Foundation

internal func bytesToHex(_ bytes: Data) -> String {
    var s = String()
    s.reserveCapacity(bytes.count * 2)
    let alphabet: [Character] = Array("0123456789abcdef")
    for byte in bytes {
        s.append(alphabet[Int(byte >> 4)])
        s.append(alphabet[Int(byte & 0x0f)])
    }
    return s
}

internal func hexToBytes(_ hex: String) -> Data {
    var s = hex
    if s.hasPrefix("0x") || s.hasPrefix("0X") {
        s = String(s.dropFirst(2))
    }
    precondition(s.count.isMultiple(of: 2), "hexToBytes: hex string must have even length, got \(s.count)")
    var out = Data(count: s.count / 2)
    let chars = Array(s)
    for i in stride(from: 0, to: chars.count, by: 2) {
        guard let hi = chars[i].hexDigitValue, let lo = chars[i + 1].hexDigitValue else {
            preconditionFailure("hexToBytes: non-hex character in \"\(hex)\"")
        }
        out[i / 2] = UInt8((hi << 4) | lo)
    }
    return out
}

internal func utf8Bytes(_ s: String) -> Data {
    return Data(s.precomposedStringWithCanonicalMapping.utf8)
}
